### PR TITLE
Assignment 2 - SK2

### PR DESCRIPTION
**Pull request by shashankkongara**

### CODE REVIEW

**Comment by julie-scherer** [comment]
This is a LLM-generated comment for `query_1.sql`: 
Hi,

Great job on your SQL query! You've successfully implemented the logic to de-duplicate the `nba_game_details` table based on the combination of `game_id`, `team_id`, and `player_id`. Your use of the `ROW_NUMBER()` window function is correct and your query is logically sound. 

Here are a few suggestions to improve your code:

1. **Readability and Comments:** Your comments are clear and explain your thought process well. However, you could improve the readability of your code by adding a comment at the top of your query explaining what it does in a concise manner. Also, consider adding comments for each major step in your query to make it easier for others to understand.

2. **Error Handling and Robustness:** Your query does not handle potential errors or edge cases. For example, what if there are null values in the `game_id`, `team_id`, or `player_id` columns? You might want to add some error handling or data validation steps to your query to ensure it can handle these situations.

3. **Modularity and Reusability:** Your query is well-structured and could be easily modified to de-duplicate other tables or based on other columns. However, you could improve the reusability of your code by parameterizing the table name and the columns to de-duplicate. This would allow you to reuse your query with different tables and columns without having to rewrite it.

4. **Ordering in Window Function:** In your window function, you've used `ORDER BY game_id`. However, since you're partitioning by `game_id`, `team_id`, and `player_id`, the `ORDER BY` clause doesn't have any effect. You can remove it to simplify your query.

Here's how you could modify your query based on these suggestions:

```sql
-- This SQL query de-duplicates the nba_game_details table based on the combination of game_id, team_id, and player_id.

-- Step 1: Assign row numbers for each unique combination of game_id, team_id, and player_id.
WITH ranked AS (
    SELECT *,
           ROW_NUMBER() OVER (PARTITION BY game_id, team_id, player_id) AS rank_num
    FROM bootcamp.nba_game_details
    WHERE game_id IS NOT NULL AND team_id IS NOT NULL AND player_id IS NOT NULL -- handle potential null values
)

-- Step 2: Select only the first row of each unique combination.
SELECT 
    game_id,
    team_id,
    team_abbreviation,
    team_city,
    player_id,
    player_name,
    nickname,
    start_position,
    comment,
    min,
    fgm,
    fga,
    fg_pct,
    fg3m,
    fg3a,
    fg3_pct,
    ftm,
    fta,
    ft_pct,
    oreb,
    dreb,
    reb,
    ast,
    stl,
    blk,
    to,
    pf,
    pts,
    plus_minus
FROM 
    ranked 
WHERE 
    rank_num = 1
```

Keep up the good work!

**Comment by julie-scherer** [comment]
This is a LLM-generated comment for `query_2.sql`: 
Hi there,

Great job on your homework submission! Your SQL statement correctly follows the assignment prompt and creates the `user_devices_cumulated` table with the correct schema. Here are some specific points of feedback:

1) **Completeness**: Your code covers all aspects of the homework prompt. You have correctly defined the table schema and included all the necessary columns (`user_id`, `browser_type`, `dates_active`, `date`). 

2) **Logical Consistency**: Your solution logically follows the given homework prompt. You have correctly created a table that will store user activity by device, which is exactly what the prompt asked for.

3) **Readability**: Your code is very readable. You have used meaningful variable names and included comments explaining what each part of the code does. This makes it easy for others to understand your thought process. The formatting of your code is also clean and consistent, which further enhances readability.

4) **Robustness and Error Handling**: As this is a DDL statement, there isn't much scope for error handling or dealing with edge cases. However, it's important to ensure that the data types of the columns match the data that will be inserted into them. Your code correctly defines the data types for each column, which should prevent errors when inserting data.

5) **Modularity and Reusability**: Your code is modular and can be easily reused to create similar tables. The comments and clear variable names make it easy to understand what each part of the code does, which would make it easier to modify or reuse this code in other contexts.

Overall, you've done an excellent job on this assignment. Keep up the good work!

**Comment by julie-scherer** [comment]
This is a LLM-generated comment for `query_3.sql`: 
Dear Student,

Great job on your SQL homework submission! Your code is well-structured, and you've done a good job of following the assignment prompt. Here are some specific points of feedback:

1) **Coverage of the Homework Prompt:** Your code covers all aspects of the homework prompt. You've correctly implemented the incremental query to populate the `user_devices_cumulated` table from the `web_events` and `devices` tables. You've also correctly used a `FULL OUTER JOIN` to ensure that data from both yesterday and today is included.

2) **Logical Flow:** Your solution logically follows the given homework prompt. The structure of your code is clear, and your use of Common Table Expressions (CTEs) makes it easy to follow the flow of data from one step to the next.

3) **Readability:** Your code is very readable. You've used meaningful variable names and included comments explaining your thought process, which makes your code easy to understand. The formatting is also clean and consistent, which further enhances readability.

4) **Robustness and Error Handling:** Your code seems robust and should handle the given inputs well. However, it's worth noting that your code doesn't include explicit error handling or checks for invalid inputs. While this wasn't a requirement for this assignment, it's a good practice to include these in your code whenever possible.

5) **Modularity and Reusability:** Your code is modular and should be relatively easy to reuse in other contexts. The use of CTEs makes it easy to isolate and understand specific parts of your code.

One suggestion for improvement would be to parameterize the dates in your query. Currently, the dates are hard-coded, which means you would need to manually update the dates each time you want to run the query for a different day. By using parameters for the dates, you could make your code more flexible and easier to reuse.

Overall, great job on this assignment! Keep up the good work.

Best,
[Your Name]

**Comment by julie-scherer** [comment]
This is a LLM-generated comment for `query_4.sql`: 
Dear Student,

Great job on your homework submission! Your code is well-structured and you've done a good job of following the assignment prompt. Here are some specific points of feedback:

1) **Coverage of the Homework Prompt:** Your code covers all aspects of the homework prompt. You've successfully converted the date list implementation into the base-2 integer datelist representation. You've also correctly used the `user_devices_cumulated` table and performed the necessary transformations.

2) **Logical Flow:** Your solution logically follows the given homework prompt. The use of Common Table Expressions (CTEs) makes your code easy to follow and understand. The transformations are done in a logical order and your comments help to explain your thought process.

3) **Readability:** Your code is very readable. You've used meaningful variable names and included comments explaining your thought process. The formatting is also clean and consistent, which makes your code easy to read.

4) **Robustness and Error Handling:** Your code seems robust and should handle the expected inputs well. However, it's not clear how your code would handle edge cases or invalid inputs. For example, what would happen if the `dates_active` array was empty or if the `user_id` or `browser_type` was null? It would be good to add some error handling or checks for these cases.

5) **Modularity and Reusability:** Your code is modular and should be easy to reuse in other contexts. The use of CTEs makes it easy to understand and modify specific parts of your code.

One thing to note is that you're using `30` in your `POW` function, while the example solution uses `31`. This might cause a difference in the results, so you might want to double-check this.

Also, you've commented out the `TO_BASE(history_int, 2) AS history_in_binary` and `BIT_COUNT(history_int, 32) AS num_days_active` lines. If these were meant to be included in your final output, you should uncomment them.

Overall, great job on this assignment! Keep up the good work.

Best,
[Your Name]

**Comment by julie-scherer** [comment]
This is a LLM-generated comment for `query_5.sql`: 
Hi there,

Great job on your homework submission! You've done a good job of following the homework prompt and your code is well-structured and easy to read. Here are some specific points of feedback:

1) **Coverage of the Homework Prompt:** Your code covers all aspects of the homework prompt. You've correctly created the `hosts_cumulated` table with the required columns and data types. You've also correctly specified the storage format as 'Parquet' and partitioned the table by 'date', as required by the prompt.

2) **Logical Flow:** Your solution logically follows the given homework prompt. The structure of your code is clear and coherent, and it's clear that you've understood the requirements of the prompt.

3) **Readability:** Your code is very readable. You've used meaningful variable names and included detailed comments explaining your thought process, which is excellent. This makes your code easy to understand for anyone who might need to read or maintain it in the future.

4) **Robustness and Error Handling:** As this is a DDL statement, there isn't much scope for error handling or dealing with edge cases. However, it's worth noting that your code is robust in the sense that it correctly creates the required table with the correct data types and partitioning.

5) **Modularity and Reusability:** Your code is modular and could easily be reused to create similar tables in the future. The comments and clear structure of your code would make it easy for someone else to adapt it for their own needs.

One minor point to note is that the example solution includes a `metric_name` and `metric_value` column, which are not included in your solution. However, these columns are not required by the homework prompt, so this is not a problem. 

Overall, great work! Keep it up.

**Comment by julie-scherer** [comment]
This is a LLM-generated comment for `query_6.sql`: 
Dear Student,

Great job on your homework submission! Your SQL query is well-structured and logically follows the given homework prompt. You've done a good job of incrementally populating the `hosts_cumulated` table from the `web_events` table. Here are some specific points of feedback:

1) **Coverage of the Homework Prompt:** Your solution covers the main aspects of the homework prompt. You've correctly used a `WITH` clause to create temporary tables for yesterday's and today's data, and then combined them to update the `hosts_cumulated` table. However, the prompt asks for a query to incrementally populate the `hosts_cumulated` table, but your solution only populates one day of data. You might want to consider a way to make your query more dynamic so it can handle multiple days of data.

2) **Logical Flow:** Your approach to the problem is logical and coherent. You've correctly used a `FULL OUTER JOIN` to ensure that data from both yesterday and today is included in the final result. This is a good approach as it ensures that all relevant data is included.

3) **Readability:** Your code is very readable. You've used meaningful variable names and included comments explaining your thought process, which makes it easy to understand what your code is doing. The overall formatting of your code is also good, which further enhances its readability.

4) **Robustness and Error Handling:** Your code seems to handle the given inputs well. However, it's not clear how your code would handle edge cases or invalid inputs. For example, what would happen if there were no events for a particular day? It would be good to include some error handling or checks to ensure that your code can handle these situations.

5) **Modularity and Reusability:** Your code is structured in a way that makes it easy to understand and reuse specific components in other contexts. The use of the `WITH` clause to create temporary tables is a good example of this. However, as mentioned earlier, you might want to consider making your query more dynamic so it can handle multiple days of data.

Overall, you've done a great job on this assignment! Keep up the good work!

Best,
[Your Name]

**Comment by julie-scherer** [comment]
This is a LLM-generated comment for `query_7.sql`: 
Hi there,

Great job on your homework submission! You've done a good job of following the homework prompt and your code is well-structured and easy to read. Here are some specific points of feedback:

1) **Coverage of the Homework Prompt:** Your code covers all aspects of the homework prompt. You've included all the required fields (`host`, `metric_name`, `metric_array`, `month_start`) in your table definition and you've also included the `partitioning` clause in the `WITH` section of your DDL statement, which matches the example solution.

2) **Logical Flow:** Your solution logically follows the given homework prompt. The comments you've included in your code make it clear what each part of your DDL statement is doing, which is very helpful for understanding your thought process.

3) **Readability:** Your code is very readable. You've used meaningful variable names and included comments explaining what each part of your code is doing. Your formatting is also clear and easy to follow.

4) **Robustness and Error Handling:** As this is a DDL statement, there isn't much scope for error handling or dealing with edge cases. However, it's worth noting that your code doesn't include any data type checks or constraints to ensure that the data being inserted into the table is of the correct type or format. This isn't a requirement for this assignment, but it's something to keep in mind for future assignments.

5) **Modularity and Reusability:** Your code is well-structured and could easily be adapted to create similar tables in the future. The comments you've included make it clear what each part of your code is doing, which would make it easier for others to understand and reuse your code.

Overall, you've done a great job on this assignment. Keep up the good work!

**Comment by julie-scherer** [comment]
This is a LLM-generated comment for `query_8.sql`: 
Overall, your solution is well-structured and logically follows the given homework prompt. You have correctly used a full outer join and handled imputing empty values in the array for windows where a host gets a visit in the middle of the array time window. However, there are a few areas where your solution could be improved:

1) **Readability:** Your code is generally well-formatted and includes meaningful variable names, which makes it easier to understand. However, your comments could be more concise and directly related to the code they are explaining. For example, the comment "-- combine metric arrays from 'yesterday' and 'today'" could be simplified to "-- Combine metric arrays". 

2) **Robustness and Error Handling:** Your code does not include any explicit error handling or checks for invalid inputs. While this may not be necessary for this specific assignment, it is a good practice to include these in your code. For example, you could add checks to ensure that the dates provided are valid and that the `metric_array` is not empty.

3) **Modularity and Reusability:** Your code is well-structured and could be easily adapted for use in other contexts. However, the hardcoded dates ('2023-08-01' and '2023-08-02') limit its reusability. Consider using variables or parameters for these dates to make your code more flexible.

4) **Completeness:** Your code appears to cover all aspects of the homework prompt. However, it's important to ensure that your code correctly handles all possible edge cases. For example, what happens if there are no records for 'today' or 'yesterday'? Your code should be able to handle these situations gracefully.

In conclusion, your solution is quite good, but there is always room for improvement. Keep up the good work!

File: README.md
Comment (HjersmanJ) [inline README.md:1]: The README should be comprehensive and give a summary on each file you have submitted.

File: submission/query_1.sql [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+WITH 
+    ranked AS (
+    -- assign row numbers for each unique combination of game_id, team_id, and player_id.
+    SELECT *,
+           ROW_NUMBER() OVER (PARTITION BY game_id, team_id, player_id ORDER BY game_id) AS rank_num
+    FROM bootcamp.nba_game_details
+)
+
+-- do not select rank_num, select remaining columns
+SELECT
```
Comment (HjersmanJ) [inline submission/query_1.sql:15]: SELECT * should suffice here.

File: submission/query_4.sql
Comment (HjersmanJ) [inline submission/query_4.sql:1]: The comments for this file are a tad too verbose. Keep the comments concise and to explain why you are performing an action.

File: submission/query_4.sql [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+-- CTE `date_list_int` calculates the binary history representation for each user and browser_type.
+date_list_int AS (
+    SELECT 
+        user_id,
+        browser_type,
+        -- Calculate the history integer using bit manipulation based on active dates.
+        CAST(SUM(
+            CASE 
+                -- If the date is active, calculate the power of 2 based on the date difference.
+                -- In TRINO, the maximum value of an INTEGER is `2^31 - 1`, so we pass `30` as constant to POW function,
```
Comment (HjersmanJ) [inline submission/query_4.sql:20]: If DATEDIFF is guaranteed to be at least 1, wouldn't 31 also work? Either way, merely a debate, not strict feedback. Good comment.

File: submission/query_4.sql [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+-- Final selection from `date_list_int` to get the user, browser, binary representation of history
+SELECT
+    *,
+    -- Convert the history integer to a binary format
+    -- TO_BASE(history_int, 2) AS history_in_binary,
+    -- Padding ensures that the binary string is always 31 bits long, with each bit corresponding to a specific day of the month
+    SUBSTRING(ARRAY_JOIN(REPEAT('0', 31), '') || TO_BASE(history_int, 2), -31) AS history_base2_string
+    -- Count the number of '1's in the binary string to determine the number of active days.
+    -- BIT_COUNT(history_int, 32) AS num_days_active
```
Comment (HjersmanJ) [inline submission/query_4.sql:41]: Commented out code should be removed from production grade code unless they contain a docstring of conditions regarding when to uncomment them.

File: submission/query_6.sql [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    WHERE
+        date = DATE('2022-12-31')
+),
+
+-- Aggregates new events by host and date for today's data from `web_events`
+today AS (
+    SELECT
+        host,
+        CAST(DATE_TRUNC('day', event_time) AS DATE) AS event_date,
+        COUNT(1)
```
Comment (HjersmanJ) [inline submission/query_6.sql:18]: We are missing `metric_name` (the 'visits').

**Review by HjersmanJ — APPROVED** [review]
Other than a few minor pieces of feedback, code is good. Remember to include a comprehensive README next time. Great job!
